EGM-certified photon IDs for 2016 data: cut based and MVA

ikrav opened the pull request:
This PR contains EGM certified cut-based and MVA photon IDs tuned on 80X samples. They are intended to be used on 2016 data. This new photon MVA uses CITK-based isolation calculation that caused several minor changes in EGM sequences that go outside of the new IDs themselves. This PR contains extra adjustments for using Tasks recently introduced in the framework rather than Sequences, although all old Sequences are kept as well for backward compatibility.
  This PR is re-submission of the PR 18348 (and a yet ealier 18013). The reviewers' comments are addressed, including the most recent comments regarding the "abs" bug, addition of new photon IDs to miniAOD production, and update of the photon ID for PuppiPhotons.

ikrav committed:
Bug fixes, added debug statements

ikrav committed:
Fixed AOD vs miniAOD detection

ikrav committed:
The first version of 80X cut-based photon ID

ikrav committed:
Added the possibility of having different egmPhotonIDSequence for AOD and miniAOD

Conflicts:
	RecoEgamma/PhotonIdentification/python/egmPhotonIDs_cff.py

ikrav committed:
Adjusted python code that deals with different VID sequences for AOD and miniAOD

ikrav committed:
New photon MVA class for 80X releases

ikrav committed:
Another adjustment for the mechanism of having different sequences in VID for AOD and miniAOD

ikrav committed:
Python configuration for the new photon MVA, imcomplete

ikrav committed:
Removed using of shower shapes from PhotonValueMapProducer that was in use until 753. Now shower shapes are taken directly from the photon object

ikrav committed:
More changes for 80X photon MVA python scripts

ikrav committed:
Further modifications to make possible to run different VID sequence for AOD and miniAOD for photons

Conflicts:
	PhysicsTools/SelectorUtils/python/tools/vid_id_tools.py

ikrav committed:
Deleted automatic production of photon MVA IDs that are too old. Left are the most recent 50ns ID (V2p1), 25ns ID (V2p1), and the new 80X ID

ikrav committed:
Fixed a problem with the order of MVA variables

ikrav committed:
Adjusted MD5 sum for the new MVA photon ID

ikrav committed:
Modified photon MVA for 80X: switched to tuning where for EE photon isolation is included in MVA after special scaling

ikrav committed:
Another adjustment of photon MVA configuration: new weights for the endcap, new working point cut values

ikrav committed:
Removed V1 of photon cut-based ID and introduced V2p2 cut-based photon ID which is a bettter tuning for 2016.

ikrav committed:
Changed the input variables rho and worst charged isolation to the form they are derived in the tuning of the latest photon MVA.

ikrav committed:
Updated pointer to indico slides for cut-based 80X photon ID

ikrav committed:
Updated MVA cut values for VID for 80X photon ID after modifying two MVA variables, as committed previously

ikrav committed:
Changes for complience with CMS coding stnadards: position of underscores, import statements, etc

ishvetso committed:
fix to run CITK isolation in VID during miniAOD production

Conflicts:
	PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py

ikrav committed:
Replaced import * with explicit imports

Conflicts:
	RecoEgamma/EgammaIsolationAlgos/python/egmPhotonIsolationAOD_cff.py

ishvetso committed:
remove redefinition of pfNoPileUpCandidates

ishvetso committed:
switch to AOD format for switchOnVIDPhotonIdProducer()

ishvetso committed:
fix possible duplications with particleFlowTmpPtrs (added ONLY when the object is not allocated)

ikrav committed:
Removed duplicated import statement

ikrav committed:
A change for PF producer that calls EGM to match the changed naming in EGM python scripts

ikrav committed:
Changes to adjust the new photon ID code to use tasks rather than sequences to match recent changes in related packages in CMSSW

ikrav committed:
A few changes to imrove python coding without change in functioning

ikrav committed:
Fixed a bug of using abs that returns an int for the photon MVA

ikrav committed:
Added the latest 80X photon IDs to the default miniAOD production

ikrav committed:
Updated the photon cut-based ID version to the EGM-recommended one for 2016+ data

cmsbuild commented:
A new Pull Request was created by  @ikrav  for master.

It involves the following packages:

CommonTools/PileupAlgos
PhysicsTools/PatAlgos
PhysicsTools/SelectorUtils
RecoEgamma/EgammaIsolationAlgos
RecoEgamma/PhotonIdentification
RecoParticleFlow/PFProducer


@perrotta, @cmsbuild, @slava77, @monttj, @davidlange6 can you please review it and eventually sign? Thanks.
@rappoccio, @gouskos, @TaiSakuma, @mmarionncern, @imarches, @ahinzmann, @acaudron, @lgray, @jdolen, @ferencek, @Sam-Harper, @rafaellopesdesa, @nhanvtran, @gkasieczka, @schoef, @JyothsnaKomaragiri, @mverzett, @cbernet, @gpetruc, @mariadalfonso, @pvmulder, @bachtis this is something you requested to watch as well.
@davidlange6 you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 78191a5a8356edc93f481bb4ec738450d69dcae7
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18645/19698/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18645/19698/summary.html

Comparison Summary:
 - You potentially added 1841 lines to the logs
 - Reco comparison results: 1704 differences found in the comparisons
 - DQMHistoTests: Total files compared: 24
 - DQMHistoTests: Total histograms compared: 1834100
 - DQMHistoTests: Total failures: 42111
 - DQMHistoTests: Total nulls: 144
 - DQMHistoTests: Total successes: 1791665
 - DQMHistoTests: Total skipped: 180
 - DQMHistoTests: Total Missing objects: 0
 - Checked 98 log files, 14 edm output root files, 24 DQM output files

slava77 commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -227,8 +227,15 @@ def miniAOD_customizeCommon(process):
+                  'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring15_50ns_nonTrig_V2p1_cff',
+                  'RecoEgamma.PhotonIdentification.Identification.cutBasedPhotonID_Spring16_V2p2_cff',
+                  'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring16_nonTrig_V1_cff']
+    switchOnVIDPhotonIdProducer(process,DataFormat.AOD, task) 
```
is this switch to DataFormat.AOD required?
Please clarify why.
Other IDs are done with .MiniAOD.

It apparently leads to some CPU increase (about 2.5% of the total miniAOD processing time).
If it's possible to avoid it without cost to physics, better use the other option.

ishvetso replied:
This means that effectively for the isolation calculation we have to use inputs as in AOD case. Of course, we have tried to use packedPFCandidates, but the problem is that at this step reducedEgamma is the input collection, and for example associatedPackedPFCandidates() is not available (at least it was failing). In CITK it is possible to use slimmedPhotons together with packedPFCandidates or gedPhotons (or reducedEgamma in this case) together with pfNoPileUpCandidates.

slava77 commented:
Notes for https://github.com/cms-sw/cmssw/pull/18645	78191a5
- code changes are in line with the description and the follow up review: the change affects photon IDs stored in miniAOD photons and also puppi MET (puppi photon selection uses updated ID)
- jenkins tests pass and comparisons with baseline show small changes only in puppi MET 
- local tests with Hgg and ZMM samples with PU show that there is a noticeable improvement in puppi MET in Hgg and essentially no change in ZMM

Here is a plot from Hgg (this PR is in red /RelValH125GGgluonfusion_13/CMSSW_9_1_0_pre3-PU25ns_91X_upgrade2017_realistic_v3-v1/GEN-SIM-RECO is used as an input; first converted to AOD)
<img width="836" alt="wf10248_metpuppi_genmetdiff" src="https://cloud.githubusercontent.com/assets/4676718/26010468/d04d90d0-3701-11e7-8a0c-2d3cf583aeb8.png">
to add some numerology: in the MET bin of 100 to 150 GeV there are almost a factor of 2 fewer events (change from 64 to 39 events).

There is a slight downside in this PR: the CPU cost is up by about 80 ms/evt or about 5% of the total miniAOD time. The increase comes from 
- a half from addition of pf[No]PileUpIso  now needed by egmPhotonIsolation
    - can the packed candidate-based definition be eventually used instead as it was done before?
- a half from additional photon isolation computation in photonIDValueMapProducer with veto cone, which again loops over all PVs and all pf candidates related to the PV.
    - this can be improved by some code improvements (e.g. compute the iso sums in the same loop)

ikrav commented:
Hi Slava,

This was done by Ivan Shvetsov in a dedicated commit:
   https://github.com/ikrav/cmssw/commit/37544411faf8abc70ef008a8a666d3677e18149f

I emailed to @ishvetso asking for an extended explanation, but as far as I recall
it was necessary after we introduced the new method (CITK-based) of computing isolations
for the new photon ID in this PR. 

Ilya 

> On May 12, 2017, at 8:10 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> @slava77 commented on this pull request.
> 
> In PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
> 
> > @@ -227,8 +227,15 @@ def miniAOD_customizeCommon(process):
>      photon_ids = ['RecoEgamma.PhotonIdentification.Identification.cutBasedPhotonID_Spring15_25ns_V1_cff',
>                    'RecoEgamma.PhotonIdentification.Identification.cutBasedPhotonID_Spring15_50ns_V1_cff',
>                    'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring15_25ns_nonTrig_V2p1_cff',
> -                  'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring15_50ns_nonTrig_V2p1_cff']
> -    switchOnVIDPhotonIdProducer(process,DataFormat.MiniAOD, task)
> +                  'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring15_50ns_nonTrig_V2p1_cff',
> +                  'RecoEgamma.PhotonIdentification.Identification.cutBasedPhotonID_Spring16_V2p2_cff',
> +                  'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring16_nonTrig_V1_cff']
> +    switchOnVIDPhotonIdProducer(process,DataFormat.AOD, task) 
> 
> is this switch to DataFormat.AOD required?
> Please clarify why.
> Other IDs are done with .MiniAOD.
> 
> It apparently leads to some CPU increase (about 2.5% of the total miniAOD processing time).
> If it's possible to avoid it without cost to physics, better use the other option.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
Hi Ilya,

I would wait for more details from Ivan before signing off.
What are the variables needed to compute photon ID which are available only in AOD?
This should be avoided to the extent possible, because otherwise we are losing the ability to rerun from miniAOD.


Slava

ikrav commented:
Hi Slava,

  What do you think about the response from Ivan? 

  How important is the ability to rerun from miniAOD? I imagine there are some authorities
out there who can say that this is required, or that this is desirable but not critical feature?
Unless you are the one.

  If it is necessary to be able to re-run from miniAOD production from miniAOD, I can imagine
we could exclude the photon MVA ID from miniAOD_tools.py, that relies on CITK, this time 
(not 100% sure that this is all it takes). However, this will be a growing issue for the future 
because EGM generally  intended to move all isolations to be CITK-based with time. 

I added several EGM conveners to the CC, but as direct emails, not sure what is
everyone’s github ID. Livia, Fabrice, Rong-Shyang, the direct link to GitHub for this
discussion is https://github.com/cms-sw/cmssw/pull/18645#discussion_r116577699

Regards,
Ilya


> On May 15, 2017, at 9:17 PM, Ivan Shvetsov <notifications@github.com> wrote:
> 
> @ishvetso commented on this pull request.
> 
> In PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
> 
> > @@ -227,8 +227,15 @@ def miniAOD_customizeCommon(process):
>      photon_ids = ['RecoEgamma.PhotonIdentification.Identification.cutBasedPhotonID_Spring15_25ns_V1_cff',
>                    'RecoEgamma.PhotonIdentification.Identification.cutBasedPhotonID_Spring15_50ns_V1_cff',
>                    'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring15_25ns_nonTrig_V2p1_cff',
> -                  'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring15_50ns_nonTrig_V2p1_cff']
> -    switchOnVIDPhotonIdProducer(process,DataFormat.MiniAOD, task)
> +                  'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring15_50ns_nonTrig_V2p1_cff',
> +                  'RecoEgamma.PhotonIdentification.Identification.cutBasedPhotonID_Spring16_V2p2_cff',
> +                  'RecoEgamma.PhotonIdentification.Identification.mvaPhotonID_Spring16_nonTrig_V1_cff']
> +    switchOnVIDPhotonIdProducer(process,DataFormat.AOD, task) 
> 
> This means that effectively for the isolation calculation we have to use inputs as in AOD case. Of course, we have tried to use packedPFCandidates, but the problem is that at this step reducedEgamma is the input collection, and for example associatedPackedPFCandidates() is not available (at least it was failing). In CITK it is possible to use slimmedPhotons together with packedPFCandidates or gedPhotons (or reducedEgamma in this case) together with pfNoPileUpCandidates.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
Hi Ilya

On 5/16/17 9:47 AM, ikrav wrote:
> 
> Hi Slava,
> 
> What do you think about the response from Ivan?

The answer from Ivan indicated that some technical problem with
implementation forced you to switch to inputs which are only in AOD.
Is that really the case, or did I misread the comment?

> 
> How important is the ability to rerun from miniAOD? I imagine there are
> some authorities
> out there who can say that this is required, or that this is desirable
> but not critical feature?

I think it's more important for the POG to be able to recompute ID from
miniAOD.
I think it's a bad choice in development to lock in and not allow
analysis to make an update to the ID from miniAOD.

This is probably more important than the fact that this change to use
AOD inputs also increases the CPU cost for miniAOD production.



> Unless you are the one.


> 
> If it is necessary to be able to re-run from miniAOD production from
> miniAOD, I can imagine
> we could exclude the photon MVA ID from miniAOD_tools.py, that relies on
> CITK, this time
> (not 100% sure that this is all it takes). However, this will be a
> growing issue for the future
> because EGM generally intended to move all isolations to be CITK-based
> with time.
> 
> I added several EGM conveners to the CC, but as direct emails, not sure
> what is
> everyone’s github ID. Livia, Fabrice, Rong-Shyang, the direct link to

you will need to add me in cc there as well to follow.


> GitHub for this
> discussion is
> https://github.com/cms-sw/cmssw/pull/18645#discussion_r116577699
> 
> Regards,
> Ilya
> 
>

ikrav commented:
Slava, so I think Ivan was not able to overcome this problem of running on miniAOD
products only, but I am not sure if he is certain it cannot be done (without significant
inefficiency or at all), or if he feels it is generally possible. I invite Ivan to comment.
Everyone is in CC now.

Ilya


> On May 16, 2017, at 8:31 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> Hi Ilya
> 
> On 5/16/17 9:47 AM, ikrav wrote:
> > 
> > Hi Slava,
> > 
> > What do you think about the response from Ivan?
> 
> The answer from Ivan indicated that some technical problem with
> implementation forced you to switch to inputs which are only in AOD.
> Is that really the case, or did I misread the comment?
> 
> > 
> > How important is the ability to rerun from miniAOD? I imagine there are
> > some authorities
> > out there who can say that this is required, or that this is desirable
> > but not critical feature?
> 
> I think it's more important for the POG to be able to recompute ID from
> miniAOD.
> I think it's a bad choice in development to lock in and not allow
> analysis to make an update to the ID from miniAOD.
> 
> This is probably more important than the fact that this change to use
> AOD inputs also increases the CPU cost for miniAOD production.
> 
> 
> 
> > Unless you are the one.
> 
> 
> > 
> > If it is necessary to be able to re-run from miniAOD production from
> > miniAOD, I can imagine
> > we could exclude the photon MVA ID from miniAOD_tools.py, that relies on
> > CITK, this time
> > (not 100% sure that this is all it takes). However, this will be a
> > growing issue for the future
> > because EGM generally intended to move all isolations to be CITK-based
> > with time.
> > 
> > I added several EGM conveners to the CC, but as direct emails, not sure
> > what is
> > everyone’s github ID. Livia, Fabrice, Rong-Shyang, the direct link to
> 
> you will need to add me in cc there as well to follow.
> 
> 
> > GitHub for this
> > discussion is
> > https://github.com/cms-sw/cmssw/pull/18645#discussion_r116577699
> > 
> > Regards,
> > Ilya
> > 
> > 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ishvetso commented:
Let me clarify the problem in the more detail. We need to remove candidates that are associated with the photon object for the footprint print removal. Ideally we would do isolation with packedPFCandidates and in this case we use associatedPackedPFCandidates() from the photon object to do footprint removal. This is not available in the reducedEgamma. So we use particleBasedIsolation map which matched photon object with candidates from particleFlow. For example, for puppi isolation in miniAOD we did it basically in the same way:

https://github.com/cms-sw/cmssw/blob/CMSSW_9_0_X/PhysicsTools/PatAlgos/python/slimming/egmIsolationsPUPPI_cfi.py#L10-L11
 the addition is that in this case we need to run pfNoPileUpCandidates. Is it really the case that it increases CPU time?

slava77 commented:
@ikrav @ishvetso 
It looks like https://github.com/cms-sw/cmssw/blob/CMSSW_9_2_0/RecoEgamma/EgammaIsolationAlgos/plugins/isolation_cones/PhotonPFIsolationWithMapBasedVeto.cc#L137 needs a modification to be able to accept a mix of reco::Photon and PackedCandidate for isolation as an input. Instead of using associatedPackedPFCandidates() (not available in reducedEgamma, as you already mentioned), use isInFootprintAlternative specialized to figure out the connection between the packedCandidate and the photon via the packed-to-pf association available from the PATPackedCandidateProducer.

This should allow to use packedCandidates instead of pfNoPileUpCandidates.

ikrav commented:
Thanks for advice, Slava, I’d need to think how to do it. Ivan would think about
it much faster, but unfortunately ( :-) ) he just graduated and they say he is on 
vacation for a week or so.

Ilya

> On May 22, 2017, at 8:43 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> @ikrav @ishvetso
> It looks like https://github.com/cms-sw/cmssw/blob/CMSSW_9_2_0/RecoEgamma/EgammaIsolationAlgos/plugins/isolation_cones/PhotonPFIsolationWithMapBasedVeto.cc#L137 needs a modification to be able to accept a mix of reco::Photon and PackedCandidate for isolation as an input. Instead of using associatedPackedPFCandidates() (not available in reducedEgamma, as you already mentioned), use isInFootprintAlternative specialized to figure out the connection between the packedCandidate and the photon via the packed-to-pf association available from the PATPackedCandidateProducer.
> 
> This should allow to use packedCandidates instead of pfNoPileUpCandidates.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ishvetso commented:
@slava77 I have a question to your comment. As far as I understand you are talking about these objects:
https://github.com/cms-sw/cmssw/blob/385cde24ef413f269d4296778539fab588e7515e/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc#L377-L378

do I understand correctly that we should 1) use particleBasedIsolation to define PF candidates that are associated with the photon object 2) matched THESE PF candidates to packedPFCandidates. Is this what you mean?

slava77 commented:
On 5/23/17 1:03 AM, Ivan Shvetsov wrote:
> @slava77 <https://github.com/slava77> I have a question to your comment.
> As far as I understand you are talking about these objects:
> https://github.com/cms-sw/cmssw/blob/385cde24ef413f269d4296778539fab588e7515e/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc#L377-L378
> 
> do I understand correctly that we should 1) use particleBasedIsolation
> to define PF candidates that are associated with the photon object 2)
> matched THESE PF candidates to packedPFCandidates. Is this what you mean?

yes: instead of
srcForIsolationCone = cms.InputTag("pfNoPileUpCandidates")
you will use packedPFCandidates together with the association between
packedPFCandidates and the particleFlow PFCandidates used to make them.

> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/18645#issuecomment-303320759>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbrKKcKd56ChVS1UiEBNOHV9zF2XWks5r8pLBgaJpZM4NVrQU>.
>

ikrav commented:
Hi Slava,

    The situation with this PR is the following: there is no issue with
MINIAODfromMINIAOD to the best of my understanding (email from me in 
another thread that you got). The only issue is the 2.5% CPU increase associated
with the use of pfNoPileUp. 
     I am wondering if it would be possible to merge this PR, and submit more optimal
code in another PR? This PR has 30 or more commits of all sorts of things, and it
has been in an open state for many months now. We keep needing to rebase, 
re-merge with other changes that are happening elsewhere but affect EGM, etc. 
    In this case, Ivan, who would be the most efficient checking if this solution being
discussed would work and implementing it, is on vacation. He responds sometimes, but 
he is not really available right now. I can give it a shot, but things usually take me longer,
since I don’t know CITK. I am worried that we would have another week or two delay until 
the new version is ready.  Or do you insist that this is resolved before merging this PR?

Thanks,
Ilya

> On May 23, 2017, at 2:29 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 5/23/17 1:03 AM, Ivan Shvetsov wrote:
> > @slava77 <https://github.com/slava77> I have a question to your comment.
> > As far as I understand you are talking about these objects:
> > https://github.com/cms-sw/cmssw/blob/385cde24ef413f269d4296778539fab588e7515e/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc#L377-L378
> > 
> > do I understand correctly that we should 1) use particleBasedIsolation
> > to define PF candidates that are associated with the photon object 2)
> > matched THESE PF candidates to packedPFCandidates. Is this what you mean?
> 
> yes: instead of
> srcForIsolationCone = cms.InputTag("pfNoPileUpCandidates")
> you will use packedPFCandidates together with the association between
> packedPFCandidates and the particleFlow PFCandidates used to make them.
> 
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/cms-sw/cmssw/pull/18645#issuecomment-303320759>, or
> > mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AEdcbrKKcKd56ChVS1UiEBNOHV9zF2XWks5r8pLBgaJpZM4NVrQU>.
> > 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ishvetso commented:
Hi all,

I totally support Ilya's suggestion. I don't feel like I could the change quickly... Thanks! 

Regards,
Ivan.

slava77 commented:
Hi Ilya

On 5/24/17 3:51 AM, ikrav wrote:
> 
> Hi Slava,
> 
> The situation with this PR is the following: there is no issue with
> MINIAODfromMINIAOD to the best of my understanding (email from me in
> another thread that you got).

I may be missing an email.
The last one I have on this topic was from May 22, where you say that
MINIAODfromMINIAOD runs
but it does not compute the photon ID. And that email ends with
questions on how to run photon ID from miniAOD.

From the context of this message, however, it looks like you were able
to run this 2016 photon ID setup from miniAOD.
Please clarify.

Thank you.

		--slava

> The only issue is the 2.5% CPU increase
> associated
> with the use of pfNoPileUp.
> I am wondering if it would be possible to merge this PR, and submit more
> optimal
> code in another PR? This PR has 30 or more commits of all sorts of
> things, and it
> has been in an open state for many months now. We keep needing to rebase,
> re-merge with other changes that are happening elsewhere but affect EGM,
> etc.
> In this case, Ivan, who would be the most efficient checking if this
> solution being
> discussed would work and implementing it, is on vacation. He responds
> sometimes, but
> he is not really available right now. I can give it a shot, but things
> usually take me longer,
> since I don’t know CITK. I am worried that we would have another week or
> two delay until
> the new version is ready. Or do you insist that this is resolved before
> merging this PR?
> 
> Thanks,
> Ilya
> 
>> On May 23, 2017, at 2:29 PM, Slava Krutelyov <notifications@github.com> wrote:
>> 
>> On 5/23/17 1:03 AM, Ivan Shvetsov wrote:
>> > @slava77 <https://github.com/slava77> I have a question to your comment.
>> > As far as I understand you are talking about these objects:
>> > https://github.com/cms-sw/cmssw/blob/385cde24ef413f269d4296778539fab588e7515e/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc#L377-L378
>> > 
>> > do I understand correctly that we should 1) use particleBasedIsolation
>> > to define PF candidates that are associated with the photon object 2)
>> > matched THESE PF candidates to packedPFCandidates. Is this what you mean?
>> 
>> yes: instead of
>> srcForIsolationCone = cms.InputTag("pfNoPileUpCandidates")
>> you will use packedPFCandidates together with the association between
>> packedPFCandidates and the particleFlow PFCandidates used to make them.
>> 
>> > 
>> > —
>> > You are receiving this because you were mentioned.
>> > Reply to this email directly, view it on GitHub
>> > <https://github.com/cms-sw/cmssw/pull/18645#issuecomment-303320759>, or
>> > mute the thread
>> > <https://github.com/notifications/unsubscribe-auth/AEdcbrKKcKd56ChVS1UiEBNOHV9zF2XWks5r8pLBgaJpZM4NVrQU>.
>> > 
>> 
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub, or mute the thread.
>> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/18645#issuecomment-303689642>, or
> mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEdcbi7jVgzz_CQXuK0u_De9J1kSey5Sks5r9Au4gaJpZM4NVrQU>.
>

ikrav commented:
Hi Slava,

  Let me try to clarify:

miniAOD_tools.py uses AOD content for photon ID. I do not see it as a problem. This script
is not used in MINIAOD-from-MINIAOD reprocessing. The script miniAOD_tools.py and
miniAOD production from AOD works fine, as one can see from the matrix tests.
The only issue is the 2.5% CPU increase, and I am asking if we can overlook this in this PR.

MINIAODfromMINIAOD reprocessing uses a different sets of scripts entirely. 
At the moment, there is a null action on the IDs there. With the this PR, nothing will
change there. VID is just not rerun. I confirmed that the new PR is not causing there
any unexpected problems by running this step (step4) interactively with the code from this PR.
If one does need to make VID rerun in this step in the future, as far as I see one can simply
configure VID in the “modify” scripts to run on the miniAOD content that. I have not
checked that, though, because I do not know how the whole system works.

Then your remark:
> From the context of this message, however, it looks like you were able
> to run this 2016 photon ID setup from miniAOD.

I am not sure exactly what you mean, so to cover all possibilities:
   - VID in an analysis job on miniAOD input runs fine, tested
   - miniAOD production from AOD runs fine, tested in matrix tests
   - miniAOD production from miniAOD runs fine, with null VID action, tested

I did have a few questions in the email from May 22 that you mention
regarding how miniAOD-from-miniAOD would run if it had a non-null action.
I am not sure if it should stop this PR, though, and I am not sure if I will get
an answer from anyone there, because the only person who really knew it is
Lindsey, and he did it awhile ago, VID developed since then in complexity.
I figure that once EGM needs this, someone will have to figure it out and make
it work anew, this PR or not.

Regards,
Ilya


> On May 25, 2017, at 12:16 AM, Slava Krutelyov <notifications@github.com> wrote:
> 
> Hi Ilya
> 
> On 5/24/17 3:51 AM, ikrav wrote:
> > 
> > Hi Slava,
> > 
> > The situation with this PR is the following: there is no issue with
> > MINIAODfromMINIAOD to the best of my understanding (email from me in
> > another thread that you got).
> 
> I may be missing an email.
> The last one I have on this topic was from May 22, where you say that
> MINIAODfromMINIAOD runs
> but it does not compute the photon ID. And that email ends with
> questions on how to run photon ID from miniAOD.
> 
> From the context of this message, however, it looks like you were able
> to run this 2016 photon ID setup from miniAOD.
> Please clarify.
> 
> Thank you.
> 
> --slava
> 
> > The only issue is the 2.5% CPU increase
> > associated
> > with the use of pfNoPileUp.
> > I am wondering if it would be possible to merge this PR, and submit more
> > optimal
> > code in another PR? This PR has 30 or more commits of all sorts of
> > things, and it
> > has been in an open state for many months now. We keep needing to rebase,
> > re-merge with other changes that are happening elsewhere but affect EGM,
> > etc.
> > In this case, Ivan, who would be the most efficient checking if this
> > solution being
> > discussed would work and implementing it, is on vacation. He responds
> > sometimes, but
> > he is not really available right now. I can give it a shot, but things
> > usually take me longer,
> > since I don’t know CITK. I am worried that we would have another week or
> > two delay until
> > the new version is ready. Or do you insist that this is resolved before
> > merging this PR?
> > 
> > Thanks,
> > Ilya
> > 
> >> On May 23, 2017, at 2:29 PM, Slava Krutelyov <notifications@github.com> wrote:
> >> 
> >> On 5/23/17 1:03 AM, Ivan Shvetsov wrote:
> >> > @slava77 <https://github.com/slava77> I have a question to your comment.
> >> > As far as I understand you are talking about these objects:
> >> > https://github.com/cms-sw/cmssw/blob/385cde24ef413f269d4296778539fab588e7515e/PhysicsTools/PatAlgos/plugins/PATPackedCandidateProducer.cc#L377-L378
> >> > 
> >> > do I understand correctly that we should 1) use particleBasedIsolation
> >> > to define PF candidates that are associated with the photon object 2)
> >> > matched THESE PF candidates to packedPFCandidates. Is this what you mean?
> >> 
> >> yes: instead of
> >> srcForIsolationCone = cms.InputTag("pfNoPileUpCandidates")
> >> you will use packedPFCandidates together with the association between
> >> packedPFCandidates and the particleFlow PFCandidates used to make them.
> >> 
> >> > 
> >> > —
> >> > You are receiving this because you were mentioned.
> >> > Reply to this email directly, view it on GitHub
> >> > <https://github.com/cms-sw/cmssw/pull/18645#issuecomment-303320759>, or
> >> > mute the thread
> >> > <https://github.com/notifications/unsubscribe-auth/AEdcbrKKcKd56ChVS1UiEBNOHV9zF2XWks5r8pLBgaJpZM4NVrQU>.
> >> > 
> >> 
> >> —
> >> You are receiving this because you were mentioned.
> >> Reply to this email directly, view it on GitHub, or mute the thread.
> >> 
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/cms-sw/cmssw/pull/18645#issuecomment-303689642>, or
> > mute the thread
> > <https://github.com/notifications/unsubscribe-auth/AEdcbi7jVgzz_CQXuK0u_De9J1kSey5Sks5r9Au4gaJpZM4NVrQU>.
> > 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
Hi Ilya

On 5/25/17 6:24 AM, ikrav wrote:
> I am not sure exactly what you mean, so to cover all possibilities:
> - VID in an analysis job on miniAOD input runs fine, tested

could you please clarify what runs fine in VID on miniAOD.
Are you able to compute these new 2016 photon IDs from baseline (old)
miniAOD
using code from this PR?

As I said earlier,
https://github.com/cms-sw/cmssw/pull/18645#issuecomment-301873748,
> This [be able to recompute the ID from miniAOD inputs] is probably more important than the fact that this change to use AOD inputs also increases the CPU cost for miniAOD production.

So, I'm looking for a clarification if this "EGM-certified photon IDs
for 2016" can be computed from existing (before this PR) miniAOD version.


> - miniAOD production from AOD runs fine, tested in matrix tests
> - miniAOD production from miniAOD runs fine, with null VID action, tested

ikrav commented:
> On May 25, 2017, at 4:27 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> Hi Ilya
> 
> On 5/25/17 6:24 AM, ikrav wrote:
> > I am not sure exactly what you mean, so to cover all possibilities:
> > - VID in an analysis job on miniAOD input runs fine, tested
> 
> could you please clarify what runs fine in VID on miniAOD.
> Are you able to compute these new 2016 photon IDs from baseline (old)
> miniAOD
> using code from this PR?

I have run the new photon ID (both cut-based and MVA) in a test analysis job
using the code from this PR on the miniAOD files from a GJet simulation sample from Spring16 MC:
   /GJet_Pt-40toInf_DoubleEMEnriched_MGG-80toInf_TuneCUETP8M1_13TeV_Pythia8/RunIISpring16MiniAODv2PUSpring16_80X_mcRun2_asymptotic_2016_miniAODv2_v0-v1/MINIAODSIM

This resulted in the expected performance (efficiencies) for signal and fakes.
This is part of my standard validation when preparing the PR.

> 
> As I said earlier,
> https://github.com/cms-sw/cmssw/pull/18645#issuecomment-301873748,
> > This [be able to recompute the ID from miniAOD inputs] is probably more important than the fact that this change to use AOD inputs also increases the CPU cost for miniAOD production.
> 
> So, I'm looking for a clarification if this "EGM-certified photon IDs
> for 2016" can be computed from existing (before this PR) miniAOD version.
> 

While I don’t see why not, I can’t tell for sure that it is possible without trying it. 
However, just as well, I cannot say for sure that it is possible with the old IDs either. 
This is just something that I have not done, and that has not been done lately by anyone, 
probably.  Well, I see from GitHub’s history that Lindsey switched MINIAOD-from-MINIAOD to 
null action on EGM objects back almost exactly a year ago.

Are you insisting that we figure out again how to do it and prove that this can be done if 
needed? I am not sure if Lindsey is available for guidance at this point. 

Cheers,
Ilya

slava77 commented:
On 5/25/17 7:40 AM, ikrav wrote:
> Are you insisting that we figure out again how to do it and prove that
> this can be done if
> needed? I am not sure if Lindsey is available for guidance at this point.

I'm still confused.
in the first part of your response:

> I have run the new photon ID (both cut-based and MVA) .. on the miniAOD files from Spring16

In the second part to the question
>  if this "EGM-certified photon IDs
>> for 2016" can be computed from existing (before this PR) miniAOD version. 

the answer is
> This is just something that I have not done

Perhaps I'm missing something subtle in the text, but as I read it now,
the first part contradicts the other.

ikrav commented:
> On May 25, 2017, at 4:52 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 5/25/17 7:40 AM, ikrav wrote:
> > Are you insisting that we figure out again how to do it and prove that
> > this can be done if
> > needed? I am not sure if Lindsey is available for guidance at this point.
> 
> I'm still confused.
> in the first part of your response:
> 
> > I have run the new photon ID (both cut-based and MVA) .. on the miniAOD files from Spring16
> 
> In the second part to the question
> > if this "EGM-certified photon IDs
> >> for 2016" can be computed from existing (before this PR) miniAOD version. 
> 
> the answer is
> > This is just something that I have not done
> 
> Perhaps I'm missing something subtle in the text, but as I read it now,
> the first part contradicts the other.
> 

Yes, sorry, indeed my response was confusing, I did not fully understand your question because
of hasty reading.

"So, I'm looking for a clarification if this "EGM-certified photon IDs for 2016" can be computed from existing (before this PR) miniAOD version.”
This is certainly true, I have done it running VID in a standard analysis job from existing miniAODs
from older Spring16 production, as described in the first part of my last email.

The only thing I am not sure about at the moment is how to make the MINIAOD-from-MINIAOD
option, step4 of production, be able to recompute VID decision, with or without this PR,
that’s what I was talking about in the second part of my email.

Sorry for being confusing. We could also have a skype chat if that helps.

Ilya

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/18645	78191a5
- code changes are in line with the description and affect products saved in miniAOD
- see notes in https://github.com/cms-sw/cmssw/pull/18645#issuecomment-301148015 for the summary of changes in performance


#18948 is created to track resolution of egmPhotonIsolation use of AOD inputs instead of a more optimal use of packedPFCandidates

davidlange6 commented on PhysicsTools/SelectorUtils/python/tools/auxiliary_cff.py:
```diff
@@ -0,0 +1,5 @@
+
+# Define constants for data formats for VID and anyone else who needs them
+class DataFormat:
```
Hi @ikrav - please give this class a name that matches the python file (and given that this python file does not define a cms sequence, its more appropriate that it not include "_cff" in its name.

davidlange6 commented on RecoEgamma/PhotonIdentification/plugins/PhotonMVAEstimatorRun2Spring16NonTrig.h:
```diff
@@ -0,0 +1,128 @@
+
+  // Define here the number and the meaning of the categories
+  // for this specific MVA
+  const int nCategories = 2;
```
why not an unsigned int - seems to recreate problems elsewhere?

davidlange6 commented on RecoEgamma/PhotonIdentification/plugins/PhotonMVAEstimatorRun2Spring16NonTrig.cc:
```diff
@@ -0,0 +1,342 @@
+  if ( std::abs(eta) < ebeeSplit)  
+    iCategory = CAT_EB;
+
+  if (std::abs(eta) >= ebeeSplit) 
```
this could just be an else, right?

ikrav commented:
Hi David,

   Why should it be unsigned int? What sort of problems does it create? I have always
thought that uint one would use when one does not fit into the “int” bit count, simply
to increase the range by x2 without increasing the allocated space. Here, however,
this is not a concern, it is a single digit const.
   Is this something specified in the CMS coding rules or such?

Thanks,
Ilya


> On May 29, 2017, at 8:54 AM, David Lange <notifications@github.com> wrote:
> 
> @davidlange6 commented on this pull request.
> 
> In RecoEgamma/PhotonIdentification/plugins/PhotonMVAEstimatorRun2Spring16NonTrig.h:
> 
> > +#include "RecoEgamma/EgammaTools/interface/EffectiveAreas.h"
> +
> +#include <vector>
> +#include <string>
> +
> +#include "TMVA/Factory.h"
> +#include "TMVA/Tools.h"
> +#include "TMVA/Reader.h"
> +
> +class PhotonMVAEstimatorRun2Spring16NonTrig : public AnyMVAEstimatorRun2Base{
> +  
> + public:
> +
> +  // Define here the number and the meaning of the categories
> +  // for this specific MVA
> +  const int nCategories = 2;
> 
> why not an unsigned int - seems to recreate problems elsewhere?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ikrav commented:
This file was meant to contain any additional python definitions/pieces used
by VID python code. Right now there is one, but this can grow. Are you saying
that we can only have one class/etc per file if we need to add more things like
DataFormat in the future? If so, I can change this to DataFormat.py, but I am not
sure if this is the most efficient way to do things keeping future in mind.

Cheers,
Ilya


> On May 29, 2017, at 8:52 AM, David Lange <notifications@github.com> wrote:
> 
> @davidlange6 commented on this pull request.
> 
> In PhysicsTools/SelectorUtils/python/tools/auxiliary_cff.py:
> 
> > @@ -0,0 +1,5 @@
> +
> +# Define constants for data formats for VID and anyone else who needs them
> +class DataFormat:
> 
> Hi @ikrav - please give this class a name that matches the python file (and given that this python file does not define a cms sequence, its more appropriate that it not include "_cff" in its name.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ikrav committed:
Several changes to follow recommendations of code reviewers

ikrav committed:
A fix to make the code compile related to the previous commit

ikrav commented:
David and all, so I introduced the three changes David requests. For that, I had to rebase because the IB release the PR branch was based on is deleted by now. I need an advice at this point: I rebased the PR branch ikrav:egm_id_80X_v3_photons_rebasedTo_CMSSW_9_2_X_2017-05-08-1100 to the latest IB release CMSSW_9_2_X_2017-05-29-1100. I did this locally on lxplus, added David's suggested changes, and pushed it to my new branch ikrav:egm_id_80X_v3_photons_rebasedTo_CMSSW_9_2_X_2017-05-29-1100. However, what do I do now? I tried to push the same content from the local area to the original branch of the PR, and failed with an error message:

error: failed to push some refs to 'git@github.com:ikrav/cmssw.git'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Merge the remote changes (e.g. 'git pull')
hint: before pushing again.

I would appreciate pointers how to deal with this short of deleting the PR and starting a new one once again.

Thanks,
Ilya

slava77 commented:
On 5/29/17 2:10 PM, ikrav wrote:
> I tried to push the same content from the local area to the original
> branch of the PR, and failed

please add "--force" to the git push command

cmsbuild commented:
Pull request #18645 was updated. @perrotta, @cmsbuild, @slava77, @monttj, @davidlange6 can you please check and sign again.

ikrav commented:
Thanks Slava, I just pushed it out.
Ilya

> On May 29, 2017, at 6:42 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 5/29/17 2:10 PM, ikrav wrote:
> > I tried to push the same content from the local area to the original
> > branch of the PR, and failed
> 
> please add "--force" to the git push command
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 618ab818e7ea65b99abc7d2313df6429f3903883

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18645/20184/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test runtestRecoEgammaPhotonIdentification had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18645/20184/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-18645/20434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D10_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D10+RecoFullGlobal_2023D10+HARVESTFullGlobal_2023D10
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-18645/23234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D18_GenSimHLBeamSpotFull14+DigiFull_2023D18+RecoFullGlobal_2023D18+HARVESTFullGlobal_2023D18
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-18645/27434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17_GenSimHLBeamSpotFull14+DigiFull_2023D17+RecoFullGlobal_2023D17+HARVESTFullGlobal_2023D17

Comparison Summary:
 - You potentially added 193 lines to the logs
 - Reco comparison results: 1721 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1786038
 - DQMHistoTests: Total failures: 28216
 - DQMHistoTests: Total nulls: 220
 - DQMHistoTests: Total successes: 1757429
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

slava77 commented:
@cmsbuild please test

the error in the unit test was apparently transient.
Note thought that the unit test in RecoEgamma/PhotonIdentification/test/runPhoton_VID.py is still using CMSSW_7_6_0_pre4 inputs.
An update to this would be useful, perhaps more so for the 92X/master version, but also for 80X as well.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 618ab818e7ea65b99abc7d2313df6429f3903883
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18645/20258/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-18645/20258/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-18645/20434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D10_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D10+RecoFullGlobal_2023D10+HARVESTFullGlobal_2023D10
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-18645/23234.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D18_GenSimHLBeamSpotFull14+DigiFull_2023D18+RecoFullGlobal_2023D18+HARVESTFullGlobal_2023D18
   * /build/cmsbld/jenkins-workarea/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-18645/27434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17_GenSimHLBeamSpotFull14+DigiFull_2023D17+RecoFullGlobal_2023D17+HARVESTFullGlobal_2023D17

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 1700 differences found in the comparisons
 - DQMHistoTests: Total files compared: 23
 - DQMHistoTests: Total histograms compared: 1786072
 - DQMHistoTests: Total failures: 23568
 - DQMHistoTests: Total nulls: 220
 - DQMHistoTests: Total successes: 1762111
 - DQMHistoTests: Total skipped: 173
 - DQMHistoTests: Total Missing objects: 0
 - Checked 94 log files, 14 edm output root files, 23 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/18645	618ab81
- updated signoff after  https://github.com/cms-sw/cmssw/pull/18645	78191a5
after mostly technical changes were introduced in a followup to comments from @davidlange6 . Comparison of pre-rebase https://github.com/cms-sw/cmssw/compare/540c89906246e1b6853e3a217950b1ced6779c42...78191a5a8356edc93f481bb4ec738450d69dcae7 and post-rebase https://github.com/cms-sw/cmssw/compare/79feef33b34112b57f68a55366b5c73a0dec7e55...618ab818e7ea65b99abc7d2313df6429f3903883 changes shows expected technical updates.
- jenkins tests pass and comparisons with baseline show changes only in PUPPI MET variables as were seen in the last version.

Also, as a reminder: #18948 is created to track resolution of egmPhotonIsolation use of AOD inputs instead of a more optimal use of packedPFCandidates

slava77 commented on RecoEgamma/PhotonIdentification/python/Identification/cutBasedPhotonID_Spring16_V2p2_cff.py:
```diff
@@ -0,0 +1,156 @@
+
+cutBasedPhotonID_Spring16_V2p2_loose.isPOGApproved = cms.untracked.bool(True)
+cutBasedPhotonID_Spring16_V2p2_medium.isPOGApproved = cms.untracked.bool(True)
+cutBasedPhotonID_Spring16_V2p2_tight.isPOGApproved = cms.untracked.bool(True)
```
the following warnings are showing up regularly now in 93X (from around 930pre1)

```
%MSG-w IdInformation:  VersionedPhotonIdProducer:egmPhotonIDs@ctor  18-Jul-2017 18:30:59 CEST pre-events
Instantiated ID: cutBasedPhotonID-Spring16-V2p2-loose
with MD5 hash: d6ce6a4f3476294bf0a3261e00170daf
This ID is not POG approved and likely under development!!!
Please make sure to report your progress with this ID at the next relevant POG meeting.

%MSG
%MSG-w IdInformation:  VersionedPhotonIdProducer:egmPhotonIDs@ctor  18-Jul-2017 18:30:59 CEST pre-events
Instantiated ID: cutBasedPhotonID-Spring16-V2p2-medium
with MD5 hash: c739cfd0b6287b8586da187c06d4053f
This ID is not POG approved and likely under development!!!
Please make sure to report your progress with this ID at the next relevant POG meeting.

%MSG
%MSG-w IdInformation:  VersionedPhotonIdProducer:egmPhotonIDs@ctor  18-Jul-2017 18:30:59 CEST pre-events
Instantiated ID: cutBasedPhotonID-Spring16-V2p2-tight
with MD5 hash: bdb623bdb1a15c13545020a919dd9530
This ID is not POG approved and likely under development!!!
Please make sure to report your progress with this ID at the next relevant POG meeting.
```
something is wrong with the hashes
